Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,509 @@
+    {
+        #if SWIFTNIO_LIBURING_DEBUG_DUMP_CQE
+        func _debugPrintCQE(_ s : String) {
+            print("Q [\(NIOThread.current)] " + s)
+        }
+
+        if count < 0 {
+            return
+        }
+
+        _debugPrintCQE(header + " CQE:s [\(cqes)] - ring flags are [\(ring.flags)]")
+        for i in 0..<count {
+            let c = cqes[i]!.pointee
+
+            let dp = io_uring_cqe_get_data(cqes[i])
+            let bitPattern : UInt = UInt(bitPattern:dp)
+            let fd = Int32(bitPattern & 0x00000000FFFFFFFF)
+            let sequenceNumber : UInt32 = UInt32((Int(bitPattern) >> 32) & 0x00FFFFFF)
+            let eventType = CqeEventType(rawValue:((Int(bitPattern) >> 32) & 0xFF000000) >> 24) // shift out the fd
+
+            let bitpatternAsPointer = UnsafeMutableRawPointer.init(bitPattern: bitPattern)
+
+            _debugPrintCQE("\(i) = fd[\(fd)] eventType[\(String(describing:eventType))] sequenceNumber[\(sequenceNumber)] res [\(c.res)] flags [\(c.flags)]  bitpattern[\(String(describing:bitpatternAsPointer))]")
+        }
+        #endif
+    }
+
+    init() {
+        cqes = UnsafeMutablePointer<UnsafeMutablePointer<io_uring_cqe>?>.allocate(capacity: Int(cqeMaxCount))
+        cqes.initialize(repeating:&emptyCqe, count:Int(cqeMaxCount))
+    }
+
+    deinit {
+        cqes.deallocate()
+    }
+
+    internal func io_uring_queue_init() throws -> () {
+        // IORING_SETUP_SQPOLL will be fundamentally useful in 5.13, as we no longer need elevated privileges as previously
+        // and CNIOLinux.io_uring_queue_init will setup a shared uring instance which allows us to use a single
+        // kernel sqpoll thread that is shared amongst all loops. NB, each process will have it's own polling thread
+        // if SQPOLL is enabled, so for hosts running multiple processes using it, some care should be taken to not overload.
+        #if SWIFTNIO_LIBURING_SQPOLL_ENABLED
+        let _sqpollEnabled : UInt32 = 1
+        #else
+        let _sqpollEnabled : UInt32 = 0
+        #endif
+        if (CNIOLinux_io_uring_queue_init(ringEntries, &ring, _sqpollEnabled ) != 0)
+         {
+             throw UringError.uringSetupFailure
+         }
+
+        _debugPrint("io_uring_queue_init \(self.ring.ring_fd)")
+     }
+
+    internal func io_uring_queue_exit() {
+        _debugPrint("io_uring_queue_exit \(self.ring.ring_fd)")
+        CNIOLinux.io_uring_queue_exit(&ring)
+    }
+
+    static internal func io_uring_use_multishot_poll() -> Bool {
+        #if SWIFTNIO_LIBURING_MULTISHOT
+        return true
+        #else
+        return false
+        #endif
+    }
+
+    // Adopting some retry code from queue.c from liburing with slight
+    // modifications - we never want to have to handle retries of
+    // SQE allocation in all places it could possibly occur.
+    // If the SQ ring is full, we may need to submit IO first
+    internal func get_sqe() -> UnsafeMutablePointer<io_uring_sqe>?
```

Comment:
```suggestion
    internal func getSQE() -> UnsafeMutablePointer<io_uring_sqe>?
```

apart from that, I think we should make this function a 

```
func withSEQ<R>(_ body: (UnsafeMutablePointer<io_uring_sqe>?) throws -> R) rethrows -> R
```

because otherwise we may need to worry about the lifetime of the pointer.